cmds/sshd/sshd.go

```diff
@@ -41,41 +45,34 @@ var (
 	dprintf = func(string, ...interface{}) {}
 )
 
-func echoCopy(w io.Writer, r io.Reader) (int64, error) {
-	var b [8192]byte
-	var err error
-	var tot int64
-	for err == nil {
-		var amt int
-		if amt, err = r.Read(b[:]); err != nil || amt < 1 {
-			fmt.Printf("Read: %v", err)
-			break
-		}
-		log.Printf("Read %d bytes: %q\n", amt, b[:amt])
-		if _, err = w.Write(b[:amt]); err != nil {
-			fmt.Printf("Write: %v", err)
-		}
-		tot += int64(amt)
-	}
-	return tot, err
-}
-
-// start a shell
+// start a command
 // TODO: use /etc/passwd, but the Go support for that is incomplete
-func runShell(c ssh.Channel, p *pty.Pty, shell string) error {
-	copy := io.Copy
+func runCommand(c ssh.Channel, p *pty.Pty, cmd string) error {
 	defer c.Close()
 
-	p.Command(shell)
-	if err := p.C.Start(); err != nil {
-		return err
-	}
-	defer p.C.Wait()
-	if *debug {
-		copy = echoCopy
+	if p != nil {
+		log.Printf("Executing PTY command %s", cmd)
+		p.Command(cmd)
+		if err := p.C.Start(); err != nil {
+			return err
+		}
+		defer p.C.Wait()
+		go io.Copy(p.Ptm, c)
+		go io.Copy(c, p.Ptm)
+	} else {
+		args := strings.Fields(cmd)
+		e := exec.Command(args[0], args[1:]...)
+		e.Stdin = c
+		e.Stderr = c
+		e.Stdout = c
+		if err := e.Start(); err != nil {
+			return err
+		}
+		log.Printf("Executing non-PTY command %s", cmd)
+		defer e.Wait()
```

Comment:
Are these defers really worth the effort? Why not just do them in line? I'm not sure I see the point and there's always the issue that you can't totally guarantee when defer happens? I note that in the open source gvisor there are a number of defers that got removed for performance, but in other cases I've seen defer removed due to its lack of precision as to when it gets run.

Reply:
Fixed, they were leftovers - I just need Process.Wait() and this is going to be moved/changed in a later PR where I add exit code handling.